### PR TITLE
Add more RFC6962 hasher tests

### PR DESCRIPTION
**Pull request by pavelkalinnikov**

### CODE REVIEW

**Comment by pavelkalinnikov** [comment]
This is to cover some tests in [ct-go/merkletree/tree_hasher_test.go](https://github.com/google/certificate-transparency-go/blob/d3b12b119dbd68296b95ccbb148ef0ca59eff3de/merkletree/tree_hasher_test.go#L82) which will be removed soon.

File: merkle/rfc6962/rfc6962_test.go [hunk 1]
```diff
@@ -39,6 +44,13 @@ func TestRfc6962Hasher(t *testing.T) {
 			want: "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
 			got:  hasher.EmptyRoot(),
 		},
+		// Check that the empty hash is not the same as the hash of an empty leaf.
+		// echo -n 00 | xxd -r -p | sha256sum
+		{
+			desc: "RFC962 Empty Leaf",
```
[the remark this answers is not in the document]
Comment (pavelkalinnikov) [reply merkle/rfc6962/rfc6962_test.go]: Done.

File: merkle/rfc6962/rfc6962_test.go [hunk 2]
```diff
@@ -52,14 +64,41 @@ func TestRfc6962Hasher(t *testing.T) {
...
+				t.Fatalf("hex.DecodeString(%x): %v", tc.want, err)
+			}
+			if got, want := tc.got, wantBytes; !bytes.Equal(got, want) {
+				t.Errorf("got %x, want %x", got, want)
+			}
+		})
+	}
+}
+
+func TestRfc6962HasherCollisions(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (pavelkalinnikov) [reply merkle/rfc6962/rfc6962_test.go]: Done.

**Comment by pavelkalinnikov** [comment]
Looks like some grpc version inconsistency between etcd and Trillian?

**Comment by daviddrysdale** [comment]
Maybe https://github.com/grpc/grpc-go/pull/2027 just landed and changes something about `ClientStream`s when we're working with etcd turned on?

**Comment by pavelkalinnikov** [comment]
Yep, looks like it, hopefully [grpc/grpc-go#/2046](https://github.com/grpc/grpc-go/pull/2046) will fix the issue.

**Comment by pavelkalinnikov** [comment]
PTAL

File: merkle/rfc6962/rfc6962_test.go [hunk 3]
```diff
@@ -52,14 +64,42 @@ func TestRfc6962Hasher(t *testing.T) {
...
+		t.Errorf("Leaf hashes should differ, but both are %x", hash1)
+	}
+
+	// Compute an intermediate subtree hash.
+	subHash1 := hasher.HashChildren(hash1, hash2)
+	// Check that this is not the same as a leaf hash of their concatenation.
+	preimage := append(hash1, hash2...)
+	forgedHash, _ := hasher.HashLeaf(preimage)
+	if bytes.Equal(subHash1, forgedHash) {
+		t.Errorf("Hasher is not second-preimage resistant")
```
[the remark this answers is not in the document]
Comment (pavelkalinnikov) [reply merkle/rfc6962/rfc6962_test.go]: As discussed in person, both wordings sound relevant according to the RFC6962 [section 2.1](https://tools.ietf.org/html/rfc6962#section-2.1), so leaving as is for now, we'll follow up if this is not correct.

**Pull request merged**